### PR TITLE
Create and join for evm

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Support create subnet and join subnet for evm.

### CODE REVIEW

File: src/manager/evm.rs [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    async fn create_subnet(&self, _from: Address, params: ConstructParams) -> Result<Address> {
+        self.ensure_same_gateway(&params.ipc_gateway_addr)?;
+
+        let min_validator_stake = params
+            .min_validator_stake
+            .atto()
+            .to_u128()
+            .ok_or_else(|| anyhow!("invalid min validator stake"))?;
+
+        log::debug!("in create subnet");
```
Comment (adlrocha) [inline src/manager/evm.rs:56]:
```suggestion
        log::debug!("calling create subnet for EVM manager");
```

File: src/manager/evm.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+        let mut route = agent_subnet_to_evm_addresses(&params.parent)?;
+        let mut root = vec![ethers::types::Address::from_str(
+            "0x0000000000000000000000000000000000000000",
+        )?];
+        root.append(&mut route);
+        log::debug!("root: {root:?}");
```
Comment (adlrocha) [inline src/manager/evm.rs:63]:
```suggestion
        log::debug!("root SubnetID as Ethereum type: {root:?}");
```

File: src/manager/evm.rs [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        let min_validator_stake = params
+            .min_validator_stake
+            .atto()
+            .to_u128()
+            .ok_or_else(|| anyhow!("invalid min validator stake"))?;
+
+        log::debug!("in create subnet");
+
+        let mut route = agent_subnet_to_evm_addresses(&params.parent)?;
+        let mut root = vec![ethers::types::Address::from_str(
```
Comment (adlrocha) [inline src/manager/evm.rs:59]: Why the first element of the route needs to be the zero address?

File: src/manager/evm.rs [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+                            // subnet_addr.to_string() returns a summary of the actual Ethereum address, not
+                            // usable in the actual code.
+                            let subnet_addr = format!("{subnet_addr:?}");
+                            log::debug!("raw subnet addr: {subnet_addr:}");
+
+                            let eth_addr = EthAddress::from_str(&subnet_addr)?;
+                            return Ok(Address::from(eth_addr));
+                        }
+                        Err(_) => {
+                            log::debug!("not of event subnet actor deployed, continue");
```
Comment (adlrocha) [inline src/manager/evm.rs:109]:
```suggestion
                            log::debug!("no event for subnet actor published yet, continue");
```

File: src/manager/evm.rs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+                            continue;
+                        }
+                    }
+                }
+                Err(anyhow!("no logs receipt"))
+            }
+            None => Err(anyhow!("no receipt to event, txn not successful")),
```
Comment (adlrocha) [inline src/manager/evm.rs:116]:
```suggestion
            None => Err(anyhow!("no receipt for event, txn not successful")),
```

File: src/manager/evm.rs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+
+        let signer = Arc::new(SignerMiddleware::new(provider, wallet));
+        let gateway_contract = Gateway::new(gateway_address, Arc::new(signer.clone()));
+        let evm_registry_contract = SubnetRegistry::new(registry_address, Arc::new(signer.clone()));
+
+        Ok(Self::new(signer, gateway_contract, evm_registry_contract))
+    }
+}
+
+/// Convert the ipc SubnetID type to an evm address. It extracts the last address from the Subnet id
```
Comment (adlrocha) [inline src/manager/evm.rs:310]: Why can't we make these two functions just one? You remove all of the childrens and then take the ones you are interested in in the relevant function? I am not a fan of functions that have a single character as the only difference in their name :)

File: src/manager/evm.rs [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+
+pub struct EthSubnetManager<M: Middleware> {
+    eth_client: Arc<M>,
+    gateway_contract: Gateway<Arc<M>>,
+    registry_contract: SubnetRegistry<Arc<M>>,
+}
+
+#[async_trait]
+impl<M: Middleware + Send + Sync + 'static> SubnetManager for EthSubnetManager<M> {
+    async fn create_subnet(&self, _from: Address, params: ConstructParams) -> Result<Address> {
```
Comment (adlrocha) [inline src/manager/evm.rs:47]: As discussed sync, we need to come up with some way from having to put the `Address` as part of the manager's signature. There is no way to choose a specific Address for the EVM manager, you can only use the default one.

File: src/server/handlers/manager/mod.rs [hunk 8]
```diff
@@ -25,9 +26,16 @@ pub mod topdown_executed;
...
-        return Err(anyhow!("Internal server error"));
+    match &subnet.config {
+        SubnetConfig::Fvm(config) => {
+            if config.auth_token.is_none() {
+                log::error!("subnet {:?} does not have auth token", subnet.id);
+                return Err(anyhow!("Internal server error"));
+            }
+        }
+        SubnetConfig::Evm(_) => {
+            // TODO: add more checks later
```
Comment (adlrocha) [inline src/server/handlers/manager/mod.rs:37]: What kind of checks are you thinking of?

**Review by adlrocha — APPROVED** [review]
LGTM overall. A few comments to address before merging, but if you have tested it and it works let's ship it :)

File: src/manager/evm.rs (hunk 7 above)
Comment (cryptoAtwill) [reply src/manager/evm.rs:47]: Yep yep, I will do the ethers wallet store then we can start using the `from`.

File: src/server/handlers/manager/mod.rs (hunk 8 above)
Comment (cryptoAtwill) [reply src/server/handlers/manager/mod.rs:37]: Ensure there are accounts configured (ofc after we have added the wallet store)?

File: src/manager/evm.rs (hunk 3 above)
Comment (cryptoAtwill) [reply src/manager/evm.rs:59]: Good one, I think I saw it before in the Solidity contracts, but now I think it's removed.

Comment (cryptoAtwill) [reply src/manager/evm.rs]: updated

**Pull request closed without merging**